fl-server/src/handlers.rs

```diff
@@ -356,7 +371,57 @@ pub async fn list_flists_handler(
     Ok(ResponseResult::Flists(flists))
 }
 
-pub async fn flist_exists(dir_path: &std::path::Path, flist_name: &String) -> io::Result<bool> {
+#[utoipa::path(
+	get,
+	path = "/v1/api/fl/preview/{flist_path}",
+	responses(
+        (status = 200, description = "Flist preview result", body = PreviewResponse),
+        (status = 400, description = "Bad request"),
+        (status = 401, description = "Unauthorized user"),
+        (status = 403, description = "Forbidden"),
+        (status = 500, description = "Internal server error"),
+	),
+    params(
+        ("flist_path" = String, Path, description = "flist file path")
+    )
+)]
+#[debug_handler]
+pub async fn preview_flist_handler(
+    Extension(cfg): Extension<config::Config>,
+    Path(flist_path): Path<String>,
+) -> impl IntoResponse {
+    let fl_path = flist_path;
+
+    match validate_flist_path(cfg.users, &cfg.flist_dir, &fl_path).await {
+        Ok(_) => (),
+        Err(err) => return Err(ResponseError::BadRequest(err.to_string())),
+    };
+
+    let content = match unpack_flist(&fl_path).await {
+        Ok(paths) => paths,
+        Err(_) => return Err(ResponseError::InternalServerError),
+    };
+
+    let bytes = match std::fs::read(&fl_path) {
+        Ok(b) => b,
+        Err(err) => {
+            log::error!(
+                "failed to read flist '{}' into bytes with error {}",
+                fl_path,
+                err
+            );
+            return Err(ResponseError::InternalServerError);
+        }
+    };
+
+    Ok(ResponseResult::PreviewFlist(PreviewResponse {
+        content,
+        metadata: cfg.store_url.join("-"),
+        checksum: sha256::digest(&bytes),
+    }))
+}
+
+async fn flist_exists(dir_path: &std::path::Path, flist_name: &String) -> io::Result<bool> {
```

Comment:
usually checking if a file exists on the filesystem or not can be done with a single syscall. Iterating over all the files in a directly to see if a file exists or not will be extremely slow (imagine there is 20k files in that dir) and unnecessary.

normally getting a "file" status on the filesystem is done with `stat` or `fstat`. this is abstracted in rust with the `std::fs::metadata`  function. Check the documentation here https://doc.rust-lang.org/std/fs/struct.Metadata.html

Reply:
If the call to this function failed, you can check the error which includes an error kind which is an enum that has `NotFound` entry. 

if a Metadata was returned, u can use that to check `is_file` or `size` etc...